### PR TITLE
USWDS - Modal: Implement behaviors and add conditional initialization

### LINKED ISSUE #5062: USWDS - Bug: Bad inits and teardowns on Uswds components when multiple instances in template in modern JS frameworks
### Describe the bug

In JS frameworks (React, Vue, Angular) when multiple instances of a dynamic uswds components are dynamically created (modals, datepickers...), calling init/on in the create lifecycle and off in the destroy lifecycle is often buggy and results in multiple inits or off() on components that are already torn down. A fix that works for the modals in Nuxt builds is presented in PR #5031, not tested in Angular or React builds.

### Steps to reproduce the bug

1. Require a dynamic JS component (example: modal) in a modern JS framework implementation (eg Nuxt)
2. init in the created lifecycle hook
3. Tear down in the destroy lifecycle hook
4. have multiple instances on that component (example: 2 modals) in the template
5. Note that the first modal will get init once as expected, the second modal will get init twice

### Expected Behavior

To use the modal example from the scenario described above, I expect each instance of modal in the template to get initiated once.

### Related code

created() {
    // initialize
    modal.on();
  }

  // remove event listeners when the component un-mounts.
  beforeDestroy() {
    modal.off();
  }

### Screenshots

_No response_

### System setup

- USWDS 3.1.0
- Windows 10
- All modern browsers (FF, Chrome, Edge)

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://designsystem.digital.gov/about/community/#community-conduct).
- [X] I checked the [current issues](https://github.com/uswds/uswds/issues?q=is%3Aissue+is%3Aopen+label%3A%22Type%3A+Bug%22) for duplicate bug reports.

### PR DESCRIPTION
**Pull request by mahoneycm**
# Summary

**Implements behaviors utility and queries DOM before initialization**. Now `modal` follows the pattern set by other dynamically created components by using the behaviors utility. Additionally, the component will query the DOM before building the component to prevent errors caused by multiple initializations being called. This is beneficial for component-level initialization.

## Breaking change

This is not a breaking change.

## Related issue

Closes #5062 

## Preview link
[Modal (federalist preview)](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-modal-behavior/iframe.html?args=&id=components-modal--default&viewMode=story)

[Test repo (Follow instructions below)](https://github.com/rachidatecs/uswds-modals)

## Problem statement

When using modern JS frameworks such as Angular or React, calling on the component to initialize multiple times caused the modal's functionality to break. The error occurred because the initialization process selects all `usa-modal` divs and rebuilds them using their ID. The newly generated div removes the ID and moves it to the wrapper. 

Once initialization is called again, it selects the new `usa-modal` div _without_ the ID, which causes build errors due to the ID being `null`

## Solution

1. During initialization, check for previously declared `modalId`before running `setUpModal` function

## Testing and review

1. Checkout this PR
2. Confirm modal js follows standard dynamic component patterns regarding behaviors util
3. Run `npm link`
4. Clone this [test repo](https://github.com/rachidatecs/uswds-modals)
5. Run `npm install` `npm run start`
6. Confirm error on the mulitple modals page
7. Return to code and run `npm link @uswds/uswds`
8. Inspect multiple modal page and confirm modals are working

## Notes

Output HTML before change:
image

After:
image

Before opening this PR, make sure you’ve done whichever of these applies to you:
- [x] Confirm that this code follows the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `git pull origin [base branch]` to pull in the most recent updates from your base and check for merge conflicts. (Often, the base branch is `develop`).
- [x] Run `npm test` and confirm that all tests pass.

### CODE REVIEW

File: packages/usa-modal/src/index.js [hunk 1]
```diff
@@ -278,51 +279,50 @@ const cleanUpModal = (baseComponent) => {
...
+
+        // this will query all openers and closers including the overlay
+        document
+          .querySelectorAll(`[aria-controls="${modalId}"]`)
+          .forEach((item) => {
+            // Turn anchor links into buttons because of
+            // VoiceOver on Safari
+            if (item.nodeName === "A") {
+              item.setAttribute("role", "button");
+              item.addEventListener("click", (e) => e.preventDefault());
```
Comment (mejiaj) [inline packages/usa-modal/src/index.js:300]: Do we need this event listener for "click" if we're calling it in line 309?

File: packages/usa-modal/src/index.js [hunk 2]
```diff
@@ -278,51 +279,50 @@ const cleanUpModal = (baseComponent) => {
...
+          });
+      });
+    },
+    teardown(root) {
+      selectOrMatches(MODAL, root).forEach((modalWindow) => {
+        cleanUpModal(modalWindow);
+        const modalId = modalWindow.id;
+
+        document
+          .querySelectorAll(`[aria-controls="${modalId}"]`)
```
Comment (mejiaj) [inline packages/usa-modal/src/index.js:319]: Again, you can use `selectOrMatches` to select and pass `root`, which could be document or something else.

File: packages/usa-modal/src/index.js [hunk 3]
```diff
@@ -278,51 +279,50 @@ const cleanUpModal = (baseComponent) => {
...
+        if (modalId) {
+          setUpModal(modalWindow);
+        }
+
+        // this will query all openers and closers including the overlay
+        document
+          .querySelectorAll(`[aria-controls="${modalId}"]`)
+          .forEach((item) => {
+            // Turn anchor links into buttons because of
+            // VoiceOver on Safari
```
Comment (mejiaj) [inline packages/usa-modal/src/index.js:297]: Is this statement correct? I thought this was for _all_ screen readers? You can also fit it all in one line.

**Review by mejiaj — COMMENTED** [review]
Thanks for updating this component. Functionality looks good. Added a few comments to potentially improve clarification. Could you also update the PR description? Not seeing links to preview or the test repo mentioned.

I've tested:
- [x] Code style (I ran prettier)
- [x] Run `npm run test`
- [x] Functionality in StorybookJS

File: packages/usa-modal/src/index.js (hunk 2 above)
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: Resolved in [1c51556](https://github.com/uswds/uswds/pull/5208/commits/1c515565af6ced9e140fe588cd754d55e483f53b)

File: packages/usa-modal/src/index.js [hunk 4]
```diff
@@ -278,51 +279,50 @@ const cleanUpModal = (baseComponent) => {
...
+    init(root) {
+      selectOrMatches(MODAL, root).forEach((modalWindow) => {
+        const modalId = modalWindow.id;
+        if (modalId) {
+          setUpModal(modalWindow);
+        }
+
+        // this will query all openers and closers including the overlay
+        document
+          .querySelectorAll(`[aria-controls="${modalId}"]`)
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: Resolved in [1c51556](https://github.com/uswds/uswds/pull/5208/commits/1c515565af6ced9e140fe588cd754d55e483f53b)

File: packages/usa-modal/src/index.js (hunk 3 above)
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: Resolved in [ee866c9](https://github.com/uswds/uswds/pull/5208/commits/ee866c9f19678cfb9b24c9161cd56f5b8f12285c)

File: packages/usa-modal/src/index.js [hunk 5]
```diff
@@ -278,51 +279,50 @@ const cleanUpModal = (baseComponent) => {
...
+  {},
+  {
+    init(root) {
+      selectOrMatches(MODAL, root).forEach((modalWindow) => {
+        const modalId = modalWindow.id;
+        if (modalId) {
+          setUpModal(modalWindow);
+        }
+
+        // this will query all openers and closers including the overlay
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: Resolved in [ee866c9](https://github.com/uswds/uswds/pull/5208/commits/ee866c9f19678cfb9b24c9161cd56f5b8f12285c)

File: packages/usa-modal/src/index.js (hunk 1 above)
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: It looks like the anchor tags get rebuilt with the modal ID as it's `href`. Removing this line will cause the page to reload with the modalId as a hash in the URL.

I think I'd like to move the comment for added clarity. Something like:

```js
          // Query all openers and closers including the overlay
          selectOrMatches(`[aria-controls="${modalId}"]`, root).forEach((modalTrigger) => {
            if (modalTrigger.nodeName === "A") {
              // Turn anchor links into buttons for screen readers
              modalTrigger.setAttribute("role", "button");

              // Prevent modal triggers from acting like links
              modalTrigger.addEventListener("click", (e) => e.preventDefault());
            }
```

Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: It looks like the anchor tags get rebuilt with the modal ID as it's `href`. Removing this line will cause the page to reload with the modalId as a hash in the URL.

I think I'd like to move the comment for added clarity. Something like:

```js
          // Query all openers and closers including the overlay
          selectOrMatches(`[aria-controls="${modalId}"]`, root).forEach((modalTrigger) => {
            if (modalTrigger.nodeName === "A") {
              // Turn anchor links into buttons for screen readers
              modalTrigger.setAttribute("role", "button");

              // Prevent modal triggers from acting like links
              modalTrigger.addEventListener("click", (e) => e.preventDefault());
            }
```

Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: Completed in [24a90b7](https://github.com/uswds/uswds/pull/5208/commits/24a90b76553b0c06ff32fa843bac2f67b28833a1)

File: packages/usa-modal/src/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: For this example, I had to move the `selectOrMatches` context from `root` to `document` in order for this to pass the unit tests. Just flagging for your awareness

[0988c22](https://github.com/uswds/uswds/pull/5208/commits/0988c227869fe850286919cee8628c73b25c539f)

**Comment by mejiaj** [comment]
@rachidatecs are you able to help us test? 

I also noticed that the initialization runs twice in your test repo. So if I have two modals, it'll run `init()` four times. Any insight as to why?

**Comment by rachidatecs** [comment]
> 

Happy to pull your branch and test the new code.

As to your question on init running twice in the test repo - that may well be the point of the branch you're on, to demo the issue. Which branch are you running?

File: packages/usa-modal/src/index.js [hunk 6]
```diff
@@ -278,51 +279,49 @@ const cleanUpModal = (baseComponent) => {
...
-};
+modal = behavior(
+  {},
+  {
+    init(root) {
+      selectOrMatches(MODAL, root).forEach((modalWindow) => {
+        const modalId = modalWindow.id;
+        if (modalId) {
+          setUpModal(modalWindow);
+        }
```
[the remark this answers is not in the document]
Comment (mahoneycm) [reply packages/usa-modal/src/index.js]: I like this better! Resolved in [76b3f21](https://github.com/uswds/uswds/pull/5208/commits/76b3f210fcc18bca67aea78cbb23b1aedfb943ca)

**Comment by mejiaj** [comment]
@rachidatecs thank you! 

For confirming the issue, I was on `master` in https://github.com/rachidatecs/uswds-modals. I couldn't find why it's firing init double the amount of times it's supposed to. 

It makes sense why the first two modals have values and the other two just have `null` for ID.

**Comment by mahoneycm** [comment]
@mejiaj That's a result of the second initialization. When it's called a second time, it selects the rebuilt `usa-modal` classes and tries to build them again.

The null values are because the _new_ `usa-modal` element doesn't have an ID like it's placeholder does before initialization

**Comment by rachidatecs** [comment]
Yep, that's actually the issue we're trying to solve. We can skirt around it but requiring the package at the page level (the app level will cause modals not on the landing page to not init, and the module level will cause those duplicate inits when more that one modal is present on a page) but that's really a workaround that avoids the issue.

**Comment by mahoneycm** [comment]
Closing in favor of #5315

**Pull request closed without merging**